# processing-in-R

LAST UPDATED: 2020-10-27

# TODO: create subfolders and group files
# TODO: add short description and usage example at the beginning of each file
# TODO: add a brief description about each script here


# subfolders:

# /plots
# /statistical_tests
# /RNA-Seq_processing


# PCA_plots_gender.R 					=>
# compare_norm_vs_vst.R					=>
# featureCounts_merge.R					=>
# get_FastQCreport.R					=>
# get_Histogram_for_features.R			=> TO 
# get_Nb_of_cases_per_group.R			=> 
# get_PicardMetrics.sh					=> TO BE CHECKED
# get_STARstats.R						=> TO BE CHEKED
# get_TranscriptByChromosome.R			=> TO BE CHECKED
# get_TrimStats_PE.sh					=> TO BE MERGE WITH _SE
# get_TrimStats_SE.sh					=> TO BE MARGED WITH _PE
# get_Venn_diagrams.R					=> NOT WORKING, TO BE CHECKED
# get_reads-distribution.R				=> TO BE CHECKED
# plot_counts_correlation.R*			=> 
# run_DESeq2_Ian.R						=> TO BE CHEKED, THEN REMOVED
# run_DESeq2_Michal.R					=> TO BE CHECKED
# run_DESeq2_plots.R					=> TO BE CHECKED
# run_DESeq2_split_chr.R				=> TO BE CHECKED AND POSSIBLY REMOVED IF NOT NEEDED
# run_DE_analysis_TAI.R					=> TO BE REMOVED
# run_Distribution_plot.R				=> TO BE CHECKED
# run_Filter_NormaliseTMM.R				=> TO BE CHECKED AND POSSIBLY REMOVED IF NOT NEEDED ANYMORE
# run_Hierarchical_clust.R				=> TO BE COMPLETED (empty file)
# run_Linear_Regression.R				=> TO BE COMPLETED (empty file)
# run_Modify_colnames.R					=> TO BE CHECKED
# run_PCA_DESeq2.R						=> PRACTICALLY FINISHED (to be cjecked)
# run_PCA_built-in.R					=> TO BE COMPLETED
# run_PCA_built-in_another.R			=> TO BE CHECKED AND POSSIBLY REMOVED IF NOT NEEDED
# run_Process_metadata.R				=> to be chcecked
# run_Switch_ENTREZID-SYMBOL.R			=> TO BE CHECKED IF IT'S WORKING CORRECTLY (possibly to be extended for other types of annotation as well)
# test_Mann-Whitney.R					=> PRACTICALLY FINSIHED (to be checked)
# test_confounding_other_methods.R		=>
# test_confounding_Mann-Whitney.R		=>
# test_confounding_correlation-coef.R	=>
# test_correlation.R					=>
