---
title: "RNA-Seq analysis - results"
output:
  html_document:
    theme: cerulean
  pdf_document: default
---

<style type="text/css">
body, td{ font-size: 18px; }
code.r{ font-size: 12px; }
pre { font-size: 12px }
</style>

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
knitr::opts_chunk$set(comment = NA)
```

This file includes RNA-seq pipelin results, including all plots and tables with stats that were outputed by each software.

### 1. Quality control on raw fastq files (FastQC)

#### 1.1 Summary

**ADD COMMENTS**

```{r,out.width="49%",out.height="20%",fig.cap="Summary plots: read1 (left), read2 (right)",fig.show='hold',fig.align='center'}

knitr::include_graphics(c("/Users/ummz/OneDrive - University of Leeds/ANALYSES/results_run_IV_Feb20/1_quality_control/postprocessed/summary_plot_R1.jpg",
                          "/Users/ummz/OneDrive - University of Leeds/ANALYSES/results_run_IV_Feb20/1_quality_control/postprocessed/summary_plot_R2.jpg"))
``` 

#### 1.2 Duplication level

**ADD COMMENTS**

```{r,out.width="49%",out.height="20%",fig.cap="Duplication level per sample: read1 (left), read2 (right)",fig.show='hold',fig.align='center'}

knitr::include_graphics(c("/Users/ummz/OneDrive - University of Leeds/ANALYSES/results_run_IV_Feb20/1_quality_control/postprocessed/duplication_level_plot_R1.jpg",
                          "/Users/ummz/OneDrive - University of Leeds/ANALYSES/results_run_IV_Feb20/1_quality_control/postprocessed/duplication_level_plot_R2.jpg"))
``` 

#### 1.3 Total reads

**ADD COMMENTS**

```{r,out.width="49%",out.height="20%",fig.cap="Total Reads: read1 (left), read2 (right)",fig.show='hold',fig.align='center'}

knitr::include_graphics("/Users/ummz/OneDrive - University of Leeds/ANALYSES/results_run_IV_Feb20/1_quality_control/postprocessed/total_reads_plot.jpg")
``` 

#### 1.4 Overrepresented sequences - BLAST

**ADD COMMENTS**

overrepresented_seqs_R1.txt
overrepresented_seqs_R2.txt



### 2. Trimming (Trimmomatic)

**create a script for step 2_trimming results to extract the info from arc_files/submit-2-trim.sh.e.00000.0**

### 3. Quality control on trimmed fastq files (FastQC)

**run FASTQC_report_R for 3_quality_control_trimmed (single-end, paired-end:paired and paired-end:unpaired)**

### 4. Read alignment (STAR)



### 5. Obtain read counts (featureCounts)


### 6. 


