---
title: "Differential expression analysis using the Swish method"
output: html_notebook
---

Differential expression analysis (on transcript- and gene-level), using the [Swish method](https://bioconductor.org/packages/release/bioc/vignettes/fishpond/inst/doc/swish.html#The_Swish_method). (it makes use of inferential replicates that are generated by Salmon)

The Swish method is based on the [samr](https://cran.r-project.org/web/packages/samr/samr.pdf)
More details can be found here: https://github.com/MikeJSeo/SAM/blob/master/sam-manual.pdf

DTE - differential transcript expression
DGE - differential gene expression

**NOTE:** the input data was generated by [Salmon](https://salmon.readthedocs.io/en/latest/salmon.html#using-salmon) which was ran on FASTA files containing trimmed reads (output from Trimmomatic). It could also be used on pre-computed alignments SAM/BAM file (output from Trimmomatic). 
`--numBootstraps 100` was set up for bootstraping. 

```{r, results='hide'}
if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
if (!requireNamespace("SummarizedExperiment", quietly = TRUE)) BiocManager::install("SummarizedExperiment")
if (!requireNamespace("fishpond", quietly = TRUE)) BiocManager::install("fishpond")
if (!requireNamespace("tximeta", quietly = TRUE)) BiocManager::install("tximeta")

library(SummarizedExperiment)
library(fishpond)
library(tximeta)
```


```{r}
# get working directory to recognise the machine
w_dir <- getwd()

# create a shortcut for the OneDrive directory where all files are stored
if(startsWith(w_dir, "/Users/michal")){           
  main_dir <- "/Users/michal/Documents/OneDrive - University of Leeds"    # on my mac
} else if (startsWith(w_dir, "/Users/ummz")) {    
  main_dir <- "/Users/ummz/Documents/OneDrive - University of Leeds"      # on uni mac    
} else {
  print("Unrecognised machine.")
}

# define wheather output files should be saved or not [TRUE / FALSE]
output_save <- FALSE
```


```{r}
# define directory with data (INPUT)
data_dir <- file.path(main_dir, "ANALYSES_archived", "run_15_Mar21")

# define directory for results (OUTPUT)
dir_out <- file.path(main_dir, "ANALYSES_archived", "run_15_Mar21", "Swish_results")
#setwd(dir_out)
```

# Use tximeta package to load Salmon outputs with inferential replicates

```{r}
# NOTE: tximeta package require coldata in order to create SummarizedExperiment object
# this file needs to be created beforhand in as tab-delimated .txt with following structure:

# samples.txt
#---------------------------------------#
#IDs	names  cond1  cond2  cond3  (...)
#11026	ID_11026  1  0  0  (...)
#11028	ID_11028  1  0  0  (...)
#11037	ID_11037  0  1  0  (...)
#---------------------------------------#

# load table data
coldata <- read.table(file.path(data_dir, "samples.txt"), header = TRUE)

# define path to annotation files (clinical / histological)
#dir_anno <- paste0(main_dir, "/data/metadata/outliers_excluded/cic_clinical_data_v2_summary_ORDERED_outliers_excluded.csv")
#dir_anno <- paste0(main_dir, "/data/metadata/outliers_excluded/slide_scores_v6_outliers_excluded.csv")

# get list of all files and add it to coldata table
coldata$files <- file.path(data_dir, "entire_output_directory", coldata$names, "quant.sf")
all(file.exists(coldata$files))

# load in the quantification data with tximeta
se <- tximeta(coldata)

# check if all the assays have been loaded
assayNames(se)

# save se object in RData format (NOT SURE IF THIS WORKED)
if(output_save == TRUE){ save(se, file = file.path(data_dir, "se.RData")) }
```

# Run differential expression analysis using Swish

# Running swish has three steps: 
# => scaling the inferential replicates, (NOT NEEDED, USED DATA GENERATED WITH length-scaledTPM)
# => labeling the rows with sufficient counts for running differential expression, 
# => and then calculating the statistics. 

# NOTE: a random seed needs to be set before running swish(), as swish makes use of pseudo-random number generation in breaking ties and in calculating permutations

# NOTE: The default number of permutations in swish is nperms=100. However, for paired datasets as this one, you may have fewer maximum permutations. In this case, there are 64 possible ways to switch the condition labels for six pairs of samples. We can set the nperms manually (or if we had just used the default value, swish would set nperms to the maximum value possible and notify the user that it had done so).

```{r}
# transcript-level
y <- se

# run analysis of two group comparison for gender (male vs. female)
y$cond1 <- factor(y$cond1, c("male","female"))

y <- scaleInfReps(y)
y <- labelKeep(y)
y <- y[mcols(y)$keep,]

set.seed(1)
y <- swish(y, x="cond1")   # don't set nperms=, use the default

# save table with results 
write.csv(mcols(y), file = file.path(dir_out, "DTE_table_gender.csv"))

# get the number of transcripts in a 5% and 1% FDR 
sink(file = file.path(dir_out, "DTE_nb_significant_gender.txt"))
cat("Numper of significant transcripts at 5%")
table(mcols(y)$qvalue < .05)   # FALSE: 86607 | TRUE: 84
cat("\n")
cat("Numper of significant transcripts at 1%")
table(mcols(y)$qvalue < .01)   # FALSE: 86613 | TRUE: 78
sink()

# get list of transcrips in a 5% FDR ("Significant transcript list")
res_DTE <- mcols(y)[mcols(y)$qvalue < .05,]

DTE_extract <- c("tx_biotype", "gene_id", "log10mean", "log2FC", "pvalue", "qvalue")
res_DTE_fin <- res_DTE[,DTE_extract]

# save data.frame with results
write.csv(res_DTE_fin, file = file.path(dir_out, "DTE_significant_list_gender.csv"))

```

```{r}
# gene-level

# summarize all of the data to the gene level
gse <- summarizeToGene(se)

gy <- gse

# run analysis of two group comparison for gender (male vs. female)
gy$cond1 <- factor(gy$cond1, c("male","female"))

gy <- scaleInfReps(gy)
gy <- labelKeep(gy)
gy <- gy[mcols(gy)$keep,]
set.seed(1)

gy <- swish(gy, x="cond1")

# save table with results 
write.csv(mcols(gy), file = file.path(dir_out, "DGE_table_gender.csv"))

# get the number of genes in a 5% and 1% FDR 
sink(file = file.path(dir_out, "DGE_nb_significant_gender.txt"))
  cat("Numper of significant genes at 5%")
  table(mcols(gy)$qvalue < .05)   # FALSE: 20778 | TRUE: 32
  cat("\n")
  cat("Numper of significant genes at 1%")
  table(mcols(gy)$qvalue < .01)   # FALSE: 20784 | TRUE: 26
sink()

# get list of genes in a 5% FDR ("Significant gene list")
res_DGE <- mcols(gy)[mcols(gy)$qvalue < .05,]
  
DGE_extract <- c("gene_name", "gene_biotype", "seq_coord_system", "description", "entrezid", "log10mean", "log2FC", "pvalue", "qvalue")
  
res_DGE_fin <- res_DGE[,DGE_extract]

# save data.frame with results
write.csv(res_DGE_fin, file = file.path(dir_out, "DGE_significant_list_gender.csv"))

```

RESULTS table columns:

**transcript-level** colnames(mcols(y))
"tx_id"              => transcript IDs (same as "tx_name")
"tx_biotype"         => transcript biotypes (29 different types)  
"tx_cds_seq_start"   => UNIMPORTANT
"tx_cds_seq_end"     => UNIMPORTANT
"gene_id"            => genes IDs
"tx_support_level"   => UNIMPORTANT (not sure what it is)
"tx_id_version"      => transcript versions (not sure what it is)
"tx_name"            => transcript names (same as "tx_id")
"log10mean"          => mean of log10
"keep"               => UNIMPORTANT
"stat"               => either a centered Wilcoxon Mann-Whitney or a signed rank statistic, aggregated over inferential replicates
"log2FC"             => a log2 fold change (the median over inferential replicates, and averaged over pairs or groups (if groups, weighted by sample size)
"pvalue"             => pvalues
"locfdr"             => local FDRs
"qvalue"             => q-values -> This is the lowest False Discovery Rate at which the gene is called significant based on the work of [John Storey](https://rss.onlinelibrary.wiley.com/doi/full/10.1111/1467-9868.00346) who invented q-values.

**gene-level** => colnames(mcols(gy))
"gene_id"            => gene IDs (analogue)
"gene_name"          => gene names (i.e. symbols, same as "symbol")
"gene_biotype"       => gene biotypes (21 different types)
"seq_coord_system"   => either "chromosome" or "scaffold"
"description"        => gene full name, including HGNC ID number (e.g. HGNC:11858)
"gene_id_version"    => gene versions (not sure what it is)
"symbol"             => gene names (i.e. symbols, same as "gene_name")
"entrezid"           => entrez IDs
"log10mean"          => mean of log10 (analogue)
"keep"               => UNIMPORTANT (analogue)
"stat"               => either a centered Wilcoxon Mann-Whitney or a signed rank statistic, aggregated over inferential replicates (analogue)
"log2FC"             => a log2 fold change (the median over inferential replicates, and averaged over pairs or groups (if groups, weighted by sample size) (analogue) 
"pvalue"             => pvalues (analogue)
"locfdr"             => local FDRs (analogue)
"qvalue"             => q-value (analogue)

Extracted from `http://statweb.stanford.edu/~tibs/SAM/sam.pdf`

**q-value** This is the lowest False Discovery Rate at which the gene is called significant based on the work of John Storey [6] who invented q-values. It is like the familiar “p-value”, adapted to the analysis of a large number of genes. The q-value measures how significant the gene is: as di > 0 increases, the corresponding q-value decreases.

**Local FDR** This is the false discovery rate for genes with scores d that fall in a window around the score for the given gene. This is in contrast to the usual FDR, which is the false discovery rate for a list of genes, whose scores exceed a given threshold. For example, if we set ∆ to a certain value, we might get upper and lower score cutpoints of ±3, yielding 100 genes with an FDR of 10%. While the local FDR for genes with scores near ±3 is probably > 10%, the local FDR for genes with the largest scores (say ±6), might be close to zero. Local false discovery rates are discussed in [3] and [1].
NOTE: In our experience, the local FDR is inherently more difficult to estimate than the usual (global) FDR. Hence, the usual FDR is a more reliable measure of the accuracy of the gene list. In particular, we use a window of at least 50 genes to estimate the local FDR at each point. This means that for the most extreme genes, the window will consist mostly of genes that are less significant than the target gene. Thus, the reported local FDR will be too large for these genes, and larger than the global FDR. The local FDR is most accurately estimated for genes near the middle of the distribution.