---
title: "Differential expression analysis using the Swish method"
output: html_notebook
---

Differential expression analysis (on transcript- and gene-level), using the [Swish method](https://bioconductor.org/packages/release/bioc/vignettes/fishpond/inst/doc/swish.html#The_Swish_method). (it makes use of inferential replicates that are generated by Salmon)

The Swish method is based on the [samr](https://cran.r-project.org/web/packages/samr/samr.pdf)
More details can be found here: https://github.com/MikeJSeo/SAM/blob/master/sam-manual.pdf

DTE - differential transcript expression
DGE - differential gene expression

**NOTE:** the input data was generated by [Salmon](https://salmon.readthedocs.io/en/latest/salmon.html#using-salmon) which was ran on FASTA files containing trimmed reads (output from Trimmomatic). It could also be used on pre-computed alignments SAM/BAM file (output from Trimmomatic). 
`--numBootstraps 100` was set up for bootstraping. 

```{r, results='hide'}
if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
if (!requireNamespace("SummarizedExperiment", quietly = TRUE)) BiocManager::install("SummarizedExperiment")
if (!requireNamespace("fishpond", quietly = TRUE)) BiocManager::install("fishpond")
if (!requireNamespace("tximeta", quietly = TRUE)) BiocManager::install("tximeta")
if (!requireNamespace("org.Hs.eg.db", quietly = TRUE)) BiocManager::install("org.Hs.eg.db")

library(SummarizedExperiment)
library(fishpond)
library(tximeta)
library(org.Hs.eg.db)
```


```{r}
# get working directory to recognise the machine
w_dir <- getwd()

# create a shortcut for the OneDrive directory where all files are stored
if(startsWith(w_dir, "/Users/michal")){           
  main_dir <- "/Users/michal/Documents/OneDrive - University of Leeds"    # on my mac
} else if (startsWith(w_dir, "/Users/ummz")) {    
  main_dir <- "/Users/ummz/Documents/OneDrive - University of Leeds"      # on uni mac    
} else {
  print("Unrecognised machine.")
}

# define wheather output files should be saved or not [TRUE / FALSE]
output_save <- TRUE
```


```{r}
# define directory with data (INPUT)
data_dir <- file.path(main_dir, "ANALYSES_archived", "run_15_Mar21")

# define directory for results (OUTPUT)
dir_out <- file.path(main_dir, "ANALYSES_archived", "run_15_Mar21", "Swish_results")
#setwd(dir_out)
```

# Use tximeta package to load Salmon outputs with inferential replicates

TABLE OF OUTPUT FILES:
# transcript-level (same set of results for gene-level)

=> first chunk
(1) table with results (whole) [.csv]
(2) number of transcripts in a 5% and 1% FDR  [.txt]
(3) list of significant transcripts [.csv]

=> plotting results 

(1) histogram of p-values [.png]
(2) number of down- and up-regulated transcripts at 5% FDR [.txt]
(3) table with 4 columns c("log10mean","log2FC","pvalue","qvalue") ordered for positive log fold change transcripts  [.csv]
(4) table with 4 columns c("log10mean","log2FC","pvalue","qvalue") ordered for negative log fold change transcripts  [.csv]
(5) - MA plot, with FDR significant transcript colored [.png]
(5bis) - MA plot as above, but with gene symbols added [.png]

```{r}
# NOTE: tximeta package require coldata in order to create SummarizedExperiment object
# this file needs to be created beforhand in as tab-delimated .txt with following structure:

# samples.txt
#---------------------------------------#
#IDs	names  cond1  cond2  cond3  (...)
#11026	ID_11026  1  0  0  (...)
#11028	ID_11028  1  0  0  (...)
#11037	ID_11037  0  1  0  (...)
#---------------------------------------#

# load table data
coldata <- read.table(file.path(data_dir, "samples.txt"), header = TRUE)

# define the feature to be used and create a new folder for results of the current run
# colnames(coldata)
# =>  "gender" | "visual_loss" | "jaw_claudication" | "ischaemic_features" | "steroids" | "age"  
run_feat <- "gender"
  
# check if directory for this feature already exists
if(dir.exists(file.path(dir_out, run_feat)) == FALSE){
  # if not, create it
  dir.create(file.path(dir_out, run_feat))
} else {
  cat("Output directory already exists")
}

# get list of all files and add it to coldata table
coldata$files <- file.path(data_dir, "entire_output_directory", coldata$names, "quant.sf")
all(file.exists(coldata$files))

# load in the quantification data with tximeta
se <- tximeta(coldata)

# check if all the assays have been loaded
assayNames(se)

# save se object in RData format (NOT SURE IF THIS WORKED)
#if(output_save == TRUE){ save(se, file = file.path(data_dir, "se.RData")) }
```

# Run differential expression analysis using Swish

Running swish has three steps: 
- scaling the inferential replicates, (NOT NEEDED, USED DATA GENERATED WITH length-scaledTPM)
- labeling the rows with sufficient counts for running differential expression, 
- and then calculating the statistics. 

**NOTE:** a random seed needs to be set before running swish(), as swish makes use of pseudo-random number generation in breaking ties and in calculating permutations

**NOTE:** The default number of permutations in swish is nperms=100. However, for paired datasets as this one, you may have fewer maximum permutations. In this case, there are 64 possible ways to switch the condition labels for six pairs of samples. We can set the nperms manually (or if we had just used the default value, swish would set nperms to the maximum value possible and notify the user that it had done so).

**List of features**
=> clinical
- gender                 male   | female
- visual_loss            no     | yes
- jaw_claudication       no     | yes
- ischaemic_features     no     | yes
- steroids               short  | long       => 'number.of.days.on.steroids.at.TAB' | short < 6 | long >= 6
- age                    young  | old        => 'age.at.BL' | young < 75 | old >= 75

=> histological  (`ID_14058` is not present in histological metadata)
- GCA_present                        no | yes 
- Giant_cells                        no | yes
- Infiltrate_around_vasa_vasorum     no | yes
- Media_destruction                  no | yes
- Neoangiogenesis                    no | yes

- Intima_pattern                     no | yes (multiple classes => modified)
- Media_pattern                      no | yes (multiple classes => modified)
- Occlusion_grade                    no | yes (multiple classes => modified)

Other histological features: (check if they could be used)
- Any.granulomatous.infiltrate.	
- Granulomatous.infiltrate.in.adventitia.	
- Granulomatous.infiltrate.in.media.	
- Granulomatous.infiltrate.in.intima.	
- Any.lymphocytic.infiltrate.	
- Lymphocytic.infiltrate.in.adventitia.	
- Lymphocytic.infiltrate.in.media.	
- Lymphocytic.infiltrate.in.intima.	
- Barcelona.score.	
- Adventitia.pattern.	
- Aggregates.	
- PALI.	
- Hyperplasia.	
- Fibrosis.	
- Oedema.	

```{r}
# transcript-level
y <- se

# set the current feature as factor 
y$gender <- factor(y$gender, c("male","female"))

# TODO: the line above needs to be variable independent, by using str(y@colData@listData), and making all level in the same format across all the features (i.e. c("male","female") to be changed to c("no","yes"))

#str(y@colData@listData)
#y@colData[, run_feat] <- factor(y@colData[, run_feat], c("male","female"))

y <- scaleInfReps(y)
y <- labelKeep(y)
y <- y[mcols(y)$keep,]

set.seed(1)
y <- swish(y, x="gender")   # don't set nperms=, use the default

# save table with results 
write.csv(mcols(y), file = file.path(dir_out, run_feat, paste0("DTE_table_", run_feat, ".csv")))

# get the number of transcripts in a 5% and 1% FDR 
sink(file = file.path(dir_out, run_feat, paste0("DTE_nb_significant_", run_feat, ".txt")))

cat("Number of significant transcripts at 5%")
table(mcols(y)$qvalue < .05)   # FALSE: 86607 | TRUE: 84
cat("\n")
cat("Number of significant transcripts at 1%")
table(mcols(y)$qvalue < .01)   # FALSE: 86613 | TRUE: 78

sink()

# get list of transcrips in a 5% FDR ("Significant transcript list")
res_DTE <- mcols(y)[mcols(y)$qvalue < .05,]

DTE_extract <- c("tx_biotype", "gene_id", "log10mean", "log2FC", "pvalue", "qvalue")
res_DTE_fin <- res_DTE[,DTE_extract]

# save data.frame with results
write.csv(res_DTE_fin, file = file.path(dir_out, run_feat, paste0("DTE_significant_list_", run_feat, ".csv")))

```


```{r}
# plotting results - transcript-level

# (1) check the distribution of p-values
png(file.path(dir_out, run_feat, paste0("DTE_histogram_pvalues_", run_feat, ".png")))
hist(mcols(y)$pvalue, col="grey")
dev.off()

# NOTE: This looks as expected for a comparison where we expect many transcripts will be affected by the treatment (IFNg stimulation of macrophage cells). There is a flat component and then an enrichment of transcripts with p-values near 0.

# (2) check which transcripts have the most extreme log2 fold change

# NOTE: it is often that many transcripts will share the same q-value, so it’s valuable to look at the log2 fold change as well (see further note below on q-value computation). 

# The log2 fold change computed by swish is the median over inferential replicates, and uses a pseudo-count of 5 on the scaled counts, to stabilize the variance on the fold change from division by small counts. 

# Here we make two vectors that give the significant genes with the lowest (most negative) and highest (most positive) log fold changes.

# (2) get the number of down- and up-regulated transcripts at 5% FDR (in terms of their log2 fold change)
sink(file = file.path(dir_out, run_feat, paste0("DTE_nb_down-up_", run_feat, ".txt")))
cat("Number of down- and up-regulated transcripts at 5% FDR (in terms of their log2 fold change)")
with(mcols(y), table(sig=qvalue < .05, sign.lfc=sign(log2FC)))
sink()

# select only significant transcripts (separately for down- and up-regulated; and ordered)
sig <- mcols(y)$qvalue < .05
lo <- order(mcols(y)$log2FC * sig)
hi <- order(-mcols(y)$log2FC * sig)

# define a list of colnames to be retained in results table
cols <- c("log10mean","log2FC","pvalue","qvalue")

# get a table with just the calculated statistics 

# (3) for the large positive log fold change transcripts (up-regulation):
top.up <- mcols(y)[hi,]
table.up <- print(as.data.frame(top.up)[,cols], digits=3)
write.csv(table.up, file = file.path(dir_out, run_feat, paste0("DTE_table_positive_log_fold_change_", run_feat, ".csv")))

# (4) for the largest negative log fold change transcripts (down-regulation):
top.down <- mcols(y)[lo,]
table.down <- print(as.data.frame(top.down)[,cols], digits=3)
write.csv(table.down, file = file.path(dir_out, run_feat, paste0("DTE_table_negative_log_fold_change_", run_feat, ".csv")))

#----------- leave it for now -----------#
#We can plot the scaled counts for the inferential replicates, and also group the samples by a covariate, in this case the cell line. The analysis was paired, so the statistic assessed if the change within pairs was consistent. Here we plot the 100th top up-regulated transcript. plotInfReps also has functionality for plotting uncertainty in single cell data, as will be covered in a later section.

#plotInfReps(y, idx=hi[100], x="condition", cov="line")
#----------------------------------------#

# (5) make an MA plot, where the transcripts in our FDR set are colored
png(file.path(dir_out, run_feat, paste0("DTE_MA_plot_", run_feat, ".png")))
plotMASwish(y, alpha=.05)
dev.off()

# (5.bis) use the addIds function from tximeta to add gene symbols. 
# By specifying gene=TRUE, this will use the gene ID to match to gene symbols for all of the transcripts.

y <- addIds(y, "SYMBOL", gene=TRUE)

# add gene symbols to MA plot
png(file.path(dir_out, run_feat, paste0("DTE_MA_plot_with_symbols", run_feat, ".png")))
plotMASwish(y, alpha=.05, xlim=c(.5,5.5))
with(
  subset(mcols(y), qvalue < .05 & abs(log2FC) > 4),
     text(log10mean, log2FC, SYMBOL,
          col="blue", pos=4, cex=.7)
)
dev.off()

```


```{r}
# gene-level

# summarize all of the data to the gene level
gse <- summarizeToGene(se)

gy <- gse

# set the current feature as factor 
gy$gender <- factor(gy$gender, c("male","female"))

gy <- scaleInfReps(gy)
gy <- labelKeep(gy)
gy <- gy[mcols(gy)$keep,]
set.seed(1)

gy <- swish(gy, x="gender")

# save table with results 
write.csv(mcols(gy), file = file.path(dir_out, run_feat, paste0("DGE_table_", run_feat, ".csv")))

# get the number of genes in a 5% and 1% FDR 
sink(file = file.path(dir_out, run_feat, paste0("DGE_nb_significant_", run_feat, ".txt")))

  cat("Number of significant genes at 5%")
  table(mcols(gy)$qvalue < .05)   # FALSE: 20778 | TRUE: 32
  cat("\n")
  cat("Number of significant genes at 1%")
  table(mcols(gy)$qvalue < .01)   # FALSE: 20784 | TRUE: 26
  
sink()

# get list of genes in a 5% FDR ("Significant gene list")
res_DGE <- mcols(gy)[mcols(gy)$qvalue < .05,]
  
DGE_extract <- c("gene_name", "gene_biotype", "seq_coord_system", "description", "entrezid", "log10mean", "log2FC", "pvalue", "qvalue")
  
res_DGE_fin <- res_DGE[,DGE_extract]

# save data.frame with results
write.csv(res_DGE_fin, file = file.path(dir_out, run_feat, paste0("DGE_significant_list_", run_feat, ".csv")))
```

```{r}
# plotting results - gene-level

# (1) check the distribution of p-values
png(file.path(dir_out, run_feat, paste0("DGE_histogram_pvalues_", run_feat, ".png")))
hist(mcols(gy)$pvalue, col="grey")
dev.off()

# NOTE: This looks as expected for a comparison where we expect many transcripts will be affected by the treatment (IFNg stimulation of macrophage cells). There is a flat component and then an enrichment of transcripts with p-values near 0.

# (2) check which genes have the most extreme log2 fold change

# (2) get the number of down- and up-regulated geness at 5% FDR (in terms of their log2 fold change)
sink(file = file.path(dir_out, run_feat, paste0("DGE_nb_down-up_", run_feat, ".txt")))
cat("Number of down- and up-regulated genes at 5% FDR (in terms of their log2 fold change)")
with(mcols(gy), table(sig=qvalue < .05, sign.lfc=sign(log2FC)))
sink()

# select only significant transcripts (separately for down- and up-regulated; and ordered)
sig <- mcols(gy)$qvalue < .05
glo <- order(mcols(gy)$log2FC * sig)
ghi <- order(-mcols(gy)$log2FC * sig)

# get a table with just the calculated statistics 

# use the same list of colnames already defined for transcripts
#cols <- c("log10mean","log2FC","pvalue","qvalue")

# (3) for the large positive log fold change transcripts (up-regulation):
gtop.up <- mcols(gy)[ghi,]
gtable.up <- print(as.data.frame(gtop.up)[,cols], digits=3)
write.csv(gtable.up, file = file.path(dir_out, run_feat, paste0("DGE_table_positive_log_fold_change_", run_feat, ".csv")))

# (4) for the largest negative log fold change transcripts (down-regulation):
gtop.down <- mcols(y)[glo,]
gtable.down <- print(as.data.frame(gtop.down)[,cols], digits=3)
write.csv(gtable.down, file = file.path(dir_out, run_feat, paste0("DGE_table_negative_log_fold_change_", run_feat, ".csv")))


#----------- leave it for now -----------#
#We can plot a particular one of these genes:

#plotInfReps(gy, idx=ghi[100], x="condition", cov="line")

#As expected, the highly up-regulated genes are involved in immune response. Many genes encoding guanylate-binding proteins (GBP) are up-regulated, and these proteins are induced by interferon, produced in response to infection by pathogenic microbes.

#We can make an MA plot, where the genes in our FDR set are colored:
#----------------------------------------#

# (5) make an MA plot, where the transcripts in our FDR set are colored
png(file.path(dir_out, run_feat, paste0("DGE_MA_plot_", run_feat, ".png")))
plotMASwish(gy, alpha=.05)
dev.off()

# (5.bis) use the addIds function from tximeta to add gene symbols. 
# By specifying gene=TRUE, this will use the gene ID to match to gene symbols for all of the transcripts.

gy <- addIds(gy, "SYMBOL", gene=TRUE)

# add gene symbols to MA plot
png(file.path(dir_out, run_feat, paste0("DGE_MA_plot_with_symbols", run_feat, ".png")))
plotMASwish(gy, alpha=.05, xlim=c(.5,5.5))
with(
  subset(mcols(gy), qvalue < .05 & abs(log2FC) > 4),
     text(log10mean, log2FC, SYMBOL,
          col="blue", pos=4, cex=.7)
)
dev.off()

```

RESULTS table columns:

**transcript-level** colnames(mcols(y))
"tx_id"              => transcript IDs (same as "tx_name")
"tx_biotype"         => transcript biotypes (29 different types)  
"tx_cds_seq_start"   => UNIMPORTANT
"tx_cds_seq_end"     => UNIMPORTANT
"gene_id"            => genes IDs
"tx_support_level"   => UNIMPORTANT (not sure what it is)
"tx_id_version"      => transcript versions (not sure what it is)
"tx_name"            => transcript names (same as "tx_id")
"log10mean"          => mean of log10
"keep"               => UNIMPORTANT
"stat"               => either a centered Wilcoxon Mann-Whitney or a signed rank statistic, aggregated over inferential replicates
"log2FC"             => a log2 fold change (the median over inferential replicates, and averaged over pairs or groups (if groups, weighted by sample size)
"pvalue"             => pvalues
"locfdr"             => local FDRs
"qvalue"             => q-values -> This is the lowest False Discovery Rate at which the gene is called significant based on the work of [John Storey](https://rss.onlinelibrary.wiley.com/doi/full/10.1111/1467-9868.00346) who invented q-values.

**gene-level** => colnames(mcols(gy))
"gene_id"            => gene IDs (analogue)
"gene_name"          => gene names (i.e. symbols, same as "symbol")
"gene_biotype"       => gene biotypes (21 different types)
"seq_coord_system"   => either "chromosome" or "scaffold"
"description"        => gene full name, including HGNC ID number (e.g. HGNC:11858)
"gene_id_version"    => gene versions (not sure what it is)
"symbol"             => gene names (i.e. symbols, same as "gene_name")
"entrezid"           => entrez IDs
"log10mean"          => mean of log10 (analogue)
"keep"               => UNIMPORTANT (analogue)
"stat"               => either a centered Wilcoxon Mann-Whitney or a signed rank statistic, aggregated over inferential replicates (analogue)
"log2FC"             => a log2 fold change (the median over inferential replicates, and averaged over pairs or groups (if groups, weighted by sample size) (analogue) 
"pvalue"             => pvalues (analogue)
"locfdr"             => local FDRs (analogue)
"qvalue"             => q-value (analogue)

Extracted from `http://statweb.stanford.edu/~tibs/SAM/sam.pdf`

**q-value** This is the lowest False Discovery Rate at which the gene is called significant based on the work of John Storey [6] who invented q-values. It is like the familiar “p-value”, adapted to the analysis of a large number of genes. The q-value measures how significant the gene is: as di > 0 increases, the corresponding q-value decreases.

**Local FDR** This is the false discovery rate for genes with scores d that fall in a window around the score for the given gene. This is in contrast to the usual FDR, which is the false discovery rate for a list of genes, whose scores exceed a given threshold. For example, if we set ∆ to a certain value, we might get upper and lower score cutpoints of ±3, yielding 100 genes with an FDR of 10%. While the local FDR for genes with scores near ±3 is probably > 10%, the local FDR for genes with the largest scores (say ±6), might be close to zero. Local false discovery rates are discussed in [3] and [1].
NOTE: In our experience, the local FDR is inherently more difficult to estimate than the usual (global) FDR. Hence, the usual FDR is a more reliable measure of the accuracy of the gene list. In particular, we use a window of at least 50 genes to estimate the local FDR at each point. This means that for the most extreme genes, the window will consist mostly of genes that are less significant than the target gene. Thus, the reported local FDR will be too large for these genes, and larger than the global FDR. The local FDR is most accurately estimated for genes near the middle of the distribution.