---
title: "Differential expression analysis using the Swish method"
output: html_notebook
---

Differential expression analysis (on transcript- and gene-level), using the [Swish method](https://bioconductor.org/packages/release/bioc/vignettes/fishpond/inst/doc/swish.html#The_Swish_method). (it makes use of inferential replicates that are generated by Salmon)

DTE - differential transcript expression
DGE - differential gene expression

**NOTE:** the input data was generated by [Salmon](https://salmon.readthedocs.io/en/latest/salmon.html#using-salmon) which was ran on FASTA files containing trimmed reads (output from Trimmomatic). It could also be used on pre-computed alignments SAM/BAM file (output from Trimmomatic). 
`--numBootstraps 100` was set up for bootstraping. 

```{r, results='hide'}
if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
if (!requireNamespace("SummarizedExperiment", quietly = TRUE)) BiocManager::install("SummarizedExperiment")
if (!requireNamespace("fishpond", quietly = TRUE)) BiocManager::install("fishpond")
if (!requireNamespace("tximeta", quietly = TRUE)) BiocManager::install("tximeta")

library(SummarizedExperiment)
library(fishpond)
library(tximeta)

```


```{r}
# get working directory to recognise the machine
w_dir <- getwd()

# create a shortcut for the OneDrive directory where all files are stored
if(startsWith(w_dir, "/Users/michal")){           
  main_dir <- "/Users/michal/Documents/OneDrive - University of Leeds"    # on my mac
} else if (startsWith(w_dir, "/Users/ummz")) {    
  main_dir <- "/Users/ummz/Documents/OneDrive - University of Leeds"      # on uni mac    
} else {
  print("Unrecognised machine.")
}

# define wheather output files should be saved or not [TRUE / FALSE]
output_save <- TRUE
```


```{r}
# define directory with data (INPUT)
data_dir <- file.path(main_dir, "ANALYSES_archived", "run_15_Mar21")

# define directory for results (OUTPUT)
dir_out <- file.path(main_dir, "ANALYSES_archived", "run_15_Mar21", "Swish_results")
#setwd(dir_out)
```

# Use tximeta package to load Salmon outputs with inferential replicates

```{r}
# NOTE: tximeta package require coldata in order to create SummarizedExperiment object
# this file needs to be created beforhand in as tab-delimated .txt with following structure:

# samples.txt
#---------------------------------------#
#IDs	names  cond1  cond2  cond3  (...)
#11026	ID_11026  1  0  0  (...)
#11028	ID_11028  1  0  0  (...)
#11037	ID_11037  0  1  0  (...)
#---------------------------------------#

# load table data
coldata <- read.table(file.path(data_dir, "samples.txt"), header = TRUE)

# get list of all files and add it to coldata table
coldata$files <- file.path(data_dir, "entire_output_directory", coldata$names, "quant.sf")
all(file.exists(coldata$files))

```







```{r}
# create sample table, which will become the column data, colData, of the final object, a SummarizedExperiment. 
# The sample table should contain all the information we need to identify the Salmon quantification directories.

files <- file.path(data_dir, "entire_output_directory", "11026__quant", "quant.sf") 
file.exists(files)

coldata <- data.frame(files, names="11026__quant", condition="A", stringsAsFactors=FALSE)
coldata

# point to a GTF file saved locally. We link the transcriptome of the Salmon index to its locally saved GTF
indexDir <- file.path(data_dir, "index", "Homo_sapiens.GRCh38_index")
fastaFTP <- c("ftp://ftp.ensembl.org/pub/release-94/fasta/homo_sapiens/cdna/Homo_sapiens.GRCh38.cdna.all.fa.gz")
gtfPath <- file.path(data_dir, "index", "Homo_sapiens.GRCh38.94.gff3.gz")
file.exists(gtfPath)

suppressPackageStartupMessages(library(tximeta))
makeLinkedTxome(indexDir=indexDir,
                source="Ensembl",
                organism="Homo Sapiens",
                release="94",
                genome="GRCh38",
                fasta=fastaFTP,
                gtf=gtfPath,
                write=FALSE)

library(tximeta)
se <- tximeta(coldata)
```










```{r}
# load table data
samples <- read.table(file.path(data_dir, "samples.txt"), header = TRUE)

# get list of all files
files <- file.path(data_dir, "entire_output_directory", samples$runs, "quant.sf")

# add IDs as names
names(files) <- c("ID_11026", "ID_11028", "ID_11037", "ID_12223", "ID_12330", "ID_12331", "ID_12388",
                  "ID_12389", "ID_12849", "ID_12855", "ID_12898", "ID_12954", "ID_14058", "ID_14455",
                  "ID_14912", "ID_14914", "ID_14915", "ID_14916", "ID_14919", "ID_14920", "ID_14922",
                  "ID_14925", "ID_14926", "ID_14927", "ID_14929", "ID_14930", "ID_14931", "ID_16082",
                  "ID_16098", "ID_16109", "ID_16111", "ID_16125", "ID_16137", "ID_16142", "ID_16153",
                  "ID_16154", "ID_16167", "ID_16187", "ID_16648", "ID_16649", "ID_8546")

# check if all files exist
all(file.exists(files))

  
txi.inf.rep <- tximport(files, type = "salmon", txOut = TRUE)

# create an assay with "counts"    "abundance" "length"    "infRep1"   "infRep2"   "infRep3", etc.

se <- SummarizedExperiment(assays=txi.inf.rep$infReps)


y_gene <- SummarizedExperiment(assays=cts_gene, colData=anno) 
#se <- SummarizedExperiment(assays=list(txi.inf.rep$counts, txi.inf.rep$abundance, txi.inf.rep$length,
#                                    txi.inf.rep$infReps$ID_11026, txi.inf.rep$infReps$ID_11028,
 #                                   txi.inf.rep$infReps$ID_11037, txi.inf.rep$infReps$ID_12223))


rownames(txi.inf.rep$infReps$ID_11026) <- rownames(txi.inf.rep$abundance)
colnames(txi.inf.rep$infReps$ID_11026) <- paste0("infRep", seq(1:100))


se <- SummarizedExperiment(assays=txi.inf.rep$infReps$ID_11026)

y <- se
y <- scaleInfReps(y) # scales counts
y <- labelKeep(y) # labels genes to keep
set.seed(1)


```

```{r}
library(macrophage)
dir <- system.file("extdata", package="macrophage")

coldata <- read.csv(file.path(dir, "coldata.csv"))

coldata <- coldata[,c(1,2,3,5)]
names(coldata) <- c("names","id","line","condition")

coldata$files <- file.path(dir, "quants", coldata$names, "quant.sf.gz")
all(file.exists(coldata$files))

se <- tximeta(coldata)


devtools::install_url("http://cran.r-project.org/src/contrib/Archive/dbplyr/dbplyr_1.3.0.tar.gz")
```





The airway experiment data package summarizes an RNA-seq experiment investigating human smooth-muscle airway cell lines treated with dexamethasone. Load the library and data set.

```{r}
library(SummarizedExperiment)
library(airway)
data(airway)
airway
```







=> use readSalmonResults() from to import with InfRep (“tximport” does not import bootstrapped)

readSalmonResults(Salmon_log = NULL, samples = NULL, directories = NULL, logExprsOffset = 1, verbose = TRUE)


